python/hsfs/core/hudi_engine.py

```diff
@@ -0,0 +1,195 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import os
+from pathlib import Path
+
+from datetime import datetime
+
+from hsfs import feature_group_commit, storage_connector
+from hsfs.core import feature_group_api, storage_connector_api
+
+
+class HudiEngine:
+    HUDI_SPARK_FORMAT = "org.apache.hudi"
+    HUDI_TABLE_NAME = "hoodie.table.name"
+    HUDI_TABLE_STORAGE_TYPE = "hoodie.datasource.write.storage.type"
+    HUDI_TABLE_OPERATION = "hoodie.datasource.write.operation"
+    HUDI_RECORD_KEY = "hoodie.datasource.write.recordkey.field"
+    HUDI_PARTITION_FIELD = "hoodie.datasource.write.partitionpath.field"
+    HUDI_PRECOMBINE_FIELD = "hoodie.datasource.write.precombine.field"
+    HUDI_HIVE_SYNC_ENABLE = "hoodie.datasource.hive_sync.enable"
+    HUDI_HIVE_SYNC_TABLE = "hoodie.datasource.hive_sync.table"
+    HUDI_HIVE_SYNC_DB = "hoodie.datasource.hive_sync.database"
+    HUDI_HIVE_SYNC_JDBC_URL = "hoodie.datasource.hive_sync.jdbcurl"
+    HUDI_HIVE_SYNC_PARTITION_FIELDS = "hoodie.datasource.hive_sync.partition_fields"
+    HUDI_KEY_GENERATOR_OPT_KEY = "hoodie.datasource.write.keygenerator.class"
+    HUDI_COMPLEX_KEY_GENERATOR_OPT_VAL = "org.apache.hudi.keygen.CustomKeyGenerator"
+    HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY = (
+        "hoodie.datasource.hive_sync.partition_extractor_class"
+    )
+    DEFAULT_HIVE_PARTITION_EXTRACTOR_CLASS_OPT_VAL = (
+        "org.apache.hudi.hive.MultiPartKeysValueExtractor"
+    )
+    HUDI_COPY_ON_WRITE = "COPY_ON_WRITE"
+    HUDI_BULK_INSERT = "bulk_insert"
+    HUDI_INSERT = "insert"
+    HUDI_UPSERT = "upsert"
+    HUDI_DELETE = "delete"
+    HUDI_QUERY_TYPE_OPT_KEY = "hoodie.datasource.query.type"
+    HUDI_QUERY_TYPE_SNAPSHOT_OPT_VAL = "snapshot"
+    HUDI_QUERY_TYPE_INCREMENTAL_OPT_VAL = "incremental"
+    HUDI_BEGIN_INSTANTTIME_OPT_KEY = "hoodie.datasource.read.begin.instanttime"
+    HUDI_END_INSTANTTIME_OPT_KEY = "hoodie.datasource.read.end.instanttime"
+    PAYLOAD_CLASS_OPT_KEY = "hoodie.datasource.write.payload.class"
+    PAYLOAD_CLASS_OPT_VAL = "org.apache.hudi.common.model.EmptyHoodieRecordPayload"
+
+    def __init__(
+        self,
+        feature_store_id,
+        feature_store_name,
+        feature_group,
+        spark_context,
+        spark_session,
+    ):
+        self._feature_group = feature_group
+        self._spark_context = spark_context
+        self._spark_session = spark_session
+        self._feature_store_id = feature_store_id
+        self._feature_store_name = feature_store_name
+        self._base_path = self._feature_group.location
+        self._table_name = feature_group.name + "_" + str(feature_group.version)
+
+        # TODO (Davit): Need to fix ComplexKeyGenerator bug
+        self._primary_key = feature_group.primary_key[0]
+        self._partition_key = feature_group.partition_key[0]
```

Comment:
Why are you considering only the first partition/primary key? In Java you build a comma separated list, I guess you should do the same here.

Reply:
done